logic/standalone/waterfall.py

```diff
@@ -0,0 +1,188 @@
+'''
+Logic module that can create waterfall objects.
+
+
+USAGE EXAMPLE:
+
+'''
+
+import logic.common.log_utils as log
+import logic.common.tiled_utils as tiled_utils
+
+#--------------------------------------------------#
+'''Variables'''
+
+FG_PREFIX = "fg"		# Prefix of fg tilelayers
+BG_PREFIX = "bg"		# Prefix of bg tilelayers
+BLOCKOUT_PREFIX = "_FALLS_"	# Prefix of tilelayers with waterfall blockout
+OBJECTGROUP_PREFIX = "objects_waterfall_"
+
+LAYER_SORT_DIFF = 10	# Difference of sort_value between consecutive tilelayers
+LAYER_SORT_ADDON = 1	# Add to sort_value
+
+
+
+
+
+#--------------------------------------------------#
+'''Public Functions'''
+
+def ScanForWaterFalls(playdo):
+	'''
+	For each applicable tilelayer:
+	 - Check layer name for sort_value used
+	 - Check layer name for template used
+	  e.g. _FALLS_meadow => uses either the "meadow bg" or "meadow fg" template object
+	 - Process the tile2D for waterfall coordinates
+
+	Then return a list of tuples that contain all of the above informations
+	  Each tuple represents a new waterfall object layer to be added
+	'''
+
+	''' =====[Pseudo-code]=====
+
+	list_waterfall_layer = []
+
+	Get all tilelayer objects
+	Get all tilelayer names
+	for tilelayer in list_all_tilelayer:
+		# Check if current layer is a viable waterfall layer
+		curr_layer_name = tiled_utils.GetNameFromObject ( tilelayer ) 
+		if curr_layer_name doesn't start with BLOCKOUT_PREFIX : continue
+
+		# Make the new sort_value string for property
+		sort_value = _CountSortInLayer( curr_layer_name, list_all_layer_names )
+
+		# Determines the template used, partly based on sort_value
+		template_name = substitute( curr_layer_name, BLOCKOUT_PREFIX, "" )
+		if sort_value starts with "fg" : template_name += " fg"
+		else : template_name += " bg"
+
+		# Process the tile2D and find the coordinates
+		list_coordinates = _MakeListOfWaterfallCoordinates( get tile2D() )
+
+		# Group all the info into a tuple, then append to the list
+		list_waterfall_layer.append( (curr_layer_name, template_name, sort_value, list_coordinates) )
+
+	return list_waterfall_layer
+	'''
+
+	return []
+
+
+
+def CreateWaterfalls(waterfall_template, playdo, list_waterfall_layer):
+	'''Read and convert each tuple in the list into a waterfalls object layer'''
+
```

Comment:
I won't review this yet, since I'll just look at ScanWaterfalls for now. But the goal was to write descriptions for the functions, not write pseudocode. The description doesn't tell me what each of the args are: